File: js/src/types/markdown-it-options.ts
import type MarkdownIt from "markdown-it";
import type { Options } from "markdown-it";
export interface MarkdownItOptions {
    "base"?: Options;
    "plugins"?: {
        [MarkdownItPluginName.SanitizeHtml]?: { [key: string]: any };
        [MarkdownItPluginName.CjkFriendly]?: { [key: string]: any };
        [MarkdownItPluginName.Emoji]?: { [key: string]: any };
        [MarkdownItPluginName.Footnote]?: { [key: string]: any };
        [MarkdownItPluginName.GitHubAlerts]?: { [key: string]: any };
        [MarkdownItPluginName.TaskCheckbox]?: { [key: string]: any };
    };
    "disabledPlugins"?: MarkdownItPluginName[];
}

export function setupMarkdownIt(markdownIt: MarkdownIt, options: MarkdownItOptions, pluginMap: Partial<Record<MarkdownItPluginName, any>>) {
    if (options.base) {
        markdownIt.set(options.base);
    }
    const plugins = options.plugins ?? {};
    const disabledPlugins = options.disabledPlugins ?? [];
    for (const pluginName of Object.keys(pluginMap) as MarkdownItPluginName[]) {
        if (!disabledPlugins.includes(pluginName)) {
            const plugin = pluginMap[pluginName];
            markdownIt.use(plugin, plugins[pluginName as keyof typeof plugins]);
        }
    }
}

export const MarkdownItPluginName = {
    SanitizeHtml: "markdown-it-sanitize-html",
    CjkFriendly: "markdown-it-cjk-friendly",
    Emoji: "markdown-it-emoji",
    Footnote: "markdown-it-footnote",
    GitHubAlerts: "markdown-it-github-alerts",
    TaskCheckbox: "markdown-it-task-checkbox",
} as const;

export type MarkdownItPluginName = typeof MarkdownItPluginName[keyof typeof MarkdownItPluginName];

File: Sources/MarkdownItShared/MarkdownItOptions.swift
//
//  MarkdownIt.swift
//  Docsloth
//
//  Created by Manabu Nakazawa on 8/9/2025.
//

import DocslothCore
import Foundation
import JavaScriptCore

public struct MarkdownItOptions: OptionsRepresentable {
    public enum Plugin: Encodable, Hashable, Sendable {
        case cjkFriendly
        case emoji
        case footnote
        case githubAlerts
        case sanitizeHTML
        case taskCheckbox
        case other(name: String)

        private static let knownPluginNames: [Self: String] = [
            .cjkFriendly: "markdown-it-cjk-friendly",
            .emoji: "markdown-it-emoji",
            .footnote: "markdown-it-footnote",
            .sanitizeHTML: "markdown-it-sanitize-html",
            .githubAlerts: "markdown-it-github-alerts",
            .taskCheckbox: "markdown-it-task-checkbox",
        ]

        public var name: String {
            switch self {
            case .cjkFriendly, .emoji, .footnote, .githubAlerts, .sanitizeHTML, .taskCheckbox:
                Self.knownPluginNames[self] ?? ""
            case let .other(name):
                name
            }
        }

        init(name: String) {
            self = Self.knownPluginNames
                .first { $0.value == name }?
                .key ?? .other(name: name)
        }

        public init(from decoder: any Decoder) throws {
            let container = try decoder.singleValueContainer()
            let name = try container.decode(String.self)
            self.init(name: name)
        }

        public func encode(to encoder: any Encoder) throws {
            var container = encoder.singleValueContainer()
            try container.encode(name)
        }
    }

    public static let `default` = Self(
        base: [
            "linkify": true,
            "breaks": true,
            "html": true,
        ],
        plugins: [:],
        disabledPlugins: []
    )

    public var base: [String: JSONValue]
    public var plugins: [Plugin: [String: JSONValue]?]
    public var disabledPlugins: [Plugin]

    public init(
        base: [String: JSONValue] = [:],
        plugins: [Plugin: [String: JSONValue]?] = [:],
        disabledPlugins: [Plugin] = [],
    ) {
        self.base = base
        self.plugins = plugins
        self.disabledPlugins = disabledPlugins
    }

    public func encode(to encoder: any Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(base, forKey: .base)
        var pluginsDict: [String: [String: JSONValue]] = [:]
        for (plugin, options) in plugins {
            pluginsDict[plugin.name] = options
        }
        try container.encode(pluginsDict, forKey: .plugins)
        try container.encode(disabledPlugins, forKey: .disabledPlugins)
    }

    private enum CodingKeys: String, CodingKey {
        case base
        case plugins
        case disabledPlugins
    }
}
